File: FBSimulatorControlTests/Tests/Unit/FBSimulatorTests.m
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import <XCTest/XCTest.h>

#import <FBControlCore/FBControlCore.h>
#import <FBSimulatorControl/FBSimulator.h>
#import <FBSimulatorControl/FBSimulatorControl-Swift.h>

#pragma mark - Test Doubles

// Stub for SimDeviceType, providing productFamilyID and screen info.
@interface _FBTestSimDeviceType : NSObject
@property (nonatomic, assign) int productFamilyID;
@property (nonatomic, assign) CGSize mainScreenSize;
@property (nonatomic, assign) float mainScreenScale;
@property (nonatomic, copy) NSString *name;
@end

@implementation _FBTestSimDeviceType
@end

// Stub for SimRuntime, providing root path.
@interface _FBTestSimRuntime : NSObject
@property (nonatomic, copy) NSString *root;
@property (nonatomic, copy) NSString *name;
@end

@implementation _FBTestSimRuntime
@end

// Stub for SimDeviceSet, providing setPath.
@interface _FBTestSimDeviceSet : NSObject
@property (nonatomic, copy) NSString *setPath;
@end

@implementation _FBTestSimDeviceSet
@end

// Comprehensive stub for SimDevice used by FBSimulator.
@interface _FBTestSimDevice : NSObject
@property (nonatomic, strong) NSUUID *UDID;
@property (nonatomic, copy) NSString *name;
@property (nonatomic, assign) NSUInteger state;
@property (nonatomic, copy) NSString *dataPath;
@property (nonatomic, strong) _FBTestSimDeviceType *deviceType;
@property (nonatomic, strong) _FBTestSimRuntime *runtime;
@property (nonatomic, strong) _FBTestSimDeviceSet *deviceSet;
@property (nonatomic, assign) mach_port_t lookupPort;
@property (nonatomic, assign) BOOL lookupShouldFail;
@end

@implementation _FBTestSimDevice

- (instancetype)init
{
  self = [super init];
  if (self) {
    _UDID = [NSUUID UUID];
    _name = @"TestSimulator";
    _state = FBiOSTargetStateBooted;
    _dataPath = @"/tmp/test-sim-data";
    _deviceType = [[_FBTestSimDeviceType alloc] init];
    _deviceType.productFamilyID = 1;
    _deviceType.mainScreenSize = CGSizeMake(750, 1334);
    _deviceType.mainScreenScale = 2.0;
    _deviceType.name = @"iPhone 14";
    _runtime = [[_FBTestSimRuntime alloc] init];
    _runtime.root = @"/Library/Developer/CoreSimulator/Profiles/Runtimes/iOS.simruntime/Contents/Resources/RuntimeRoot";
    _runtime.name = @"iOS 17.0";
    _deviceSet = [[_FBTestSimDeviceSet alloc] init];
    _deviceSet.setPath = [@"~/Library/Developer/CoreSimulator/Devices" stringByExpandingTildeInPath];
    _lookupPort = MACH_PORT_NULL;
    _lookupShouldFail = NO;
  }
  return self;
}

- (mach_port_t)lookup:(NSString *)name error:(NSError **)error
{
  if (_lookupShouldFail) {
    if (error) {
      *error = [NSError errorWithDomain:@"FBSimulatorTestDomain" code:1 userInfo:@{NSLocalizedDescriptionKey : @"Port not found"}];
    }
    return MACH_PORT_NULL;
  }
  return _lookupPort;
}

@end

// Minimal event reporter stub.
@interface _FBTestEventReporter : NSObject
@end

@implementation _FBTestEventReporter
@end

#pragma mark - Helper

static FBSimulator *_createSimulatorWithDevice(_FBTestSimDevice *device)
{
  id<FBControlCoreLogger> logger = [FBControlCoreLoggerFactory loggerToConsumer:[FBNullDataConsumer new]];
  id reporter = [_FBTestEventReporter new];
  return [[FBSimulator alloc] initWithDevice:(id)device
                               configuration:FBSimulatorConfiguration.defaultConfiguration
                                         set:nil
                          auxillaryDirectory:@"/tmp/test-aux"
                                      logger:logger
                                    reporter:reporter];
}

#pragma mark - Test Class

@interface FBSimulatorTests : XCTestCase
@end

@implementation FBSimulatorTests
{
  _FBTestSimDevice *_stubDevice;
  FBSimulator *_simulator;
}

- (void)setUp
{
  [super setUp];
  _stubDevice = [[_FBTestSimDevice alloc] init];
  _simulator = _createSimulatorWithDevice(_stubDevice);
}

#pragma mark - Product Family

- (void)testProductFamily_WhenFamilyIDIs1_ReturnsiPhone
{
  _stubDevice.deviceType.productFamilyID = 1;
  FBSimulator *sim = _createSimulatorWithDevice(_stubDevice);
  XCTAssertEqual(
    sim.productFamily,
    FBControlCoreProductFamilyiPhone,
    @"productFamilyID 1 should map to iPhone"
  );
}

- (void)testProductFamily_WhenFamilyIDIs2_ReturnsiPad
{
  _stubDevice.deviceType.productFamilyID = 2;
  FBSimulator *sim = _createSimulatorWithDevice(_stubDevice);
  XCTAssertEqual(
    sim.productFamily,
    FBControlCoreProductFamilyiPad,
    @"productFamilyID 2 should map to iPad"
  );
}

- (void)testProductFamily_WhenFamilyIDIs3_ReturnsAppleTV
{
  _stubDevice.deviceType.productFamilyID = 3;
  FBSimulator *sim = _createSimulatorWithDevice(_stubDevice);
  XCTAssertEqual(
    sim.productFamily,
    FBControlCoreProductFamilyAppleTV,
    @"productFamilyID 3 should map to AppleTV"
  );
}

- (void)testProductFamily_WhenFamilyIDIs4_ReturnsAppleWatch
{
  _stubDevice.deviceType.productFamilyID = 4;
  FBSimulator *sim = _createSimulatorWithDevice(_stubDevice);
  XCTAssertEqual(
    sim.productFamily,
    FBControlCoreProductFamilyAppleWatch,
    @"productFamilyID 4 should map to AppleWatch"
  );
}

- (void)testProductFamily_WhenFamilyIDIsUnknown_ReturnsUnknown
{
  _stubDevice.deviceType.productFamilyID = 99;
  FBSimulator *sim = _createSimulatorWithDevice(_stubDevice);
  XCTAssertEqual(
    sim.productFamily,
    FBControlCoreProductFamilyUnknown,
    @"Unrecognized productFamilyID should map to Unknown"
  );
}

#pragma mark - Custom Device Set Path

- (void)testCustomDeviceSetPath_WhenDefaultPath_ReturnsNil
{
  _stubDevice.deviceSet.setPath = [@"~/Library/Developer/CoreSimulator/Devices" stringByExpandingTildeInPath];
  FBSimulator *sim = _createSimulatorWithDevice(_stubDevice);
  XCTAssertNil(
    sim.customDeviceSetPath,
    @"customDeviceSetPath should be nil when using the default device set path"
  );
}

- (void)testCustomDeviceSetPath_WhenCustomPath_ReturnsPath
{
  NSString *customPath = @"/custom/simulator/devices";
  _stubDevice.deviceSet.setPath = customPath;
  FBSimulator *sim = _createSimulatorWithDevice(_stubDevice);
  XCTAssertEqualObjects(
    sim.customDeviceSetPath,
    customPath,
    @"customDeviceSetPath should return the custom path when not using the default"
  );
}

#pragma mark - Auxiliary Directory From SimDevice

- (void)testAuxillaryDirectoryFromSimDevice_AppendsSubdirectory
{
  _stubDevice.dataPath = @"/path/to/sim/data";
  NSString *expected = @"/path/to/sim/data/fbsimulatorcontrol";
  // The auxillaryDirectory was set in setUp via the init parameter, so test the class method indirectly
  // by creating a simulator using the convenience init that calls auxillaryDirectoryFromSimDevice:
  id<FBControlCoreLogger> logger = [FBControlCoreLoggerFactory loggerToConsumer:[FBNullDataConsumer new]];
  id reporter = [_FBTestEventReporter new];
  FBSimulator *simViaConvenienceInit = [[FBSimulator alloc] initWithDevice:(id)_stubDevice
                                                                    logger:logger
                                                                  reporter:reporter];
  XCTAssertEqualObjects(
    simViaConvenienceInit.auxillaryDirectory,
    expected,
    @"auxillaryDirectory should be dataPath + /fbsimulatorcontrol"
  );
}

#pragma mark - Screen Info

- (void)testScreenInfo_ReturnsDeviceTypeScreenDimensions
{
  _stubDevice.deviceType.mainScreenSize = CGSizeMake(1170, 2532);
  _stubDevice.deviceType.mainScreenScale = 3.0f;
  FBSimulator *sim = _createSimulatorWithDevice(_stubDevice);
  FBiOSTargetScreenInfo *screenInfo = sim.screenInfo;
  XCTAssertNotNil(screenInfo, @"screenInfo should not be nil");
  XCTAssertEqual(
    screenInfo.widthPixels,
    (NSUInteger)1170,
    @"widthPixels should match device type mainScreenSize.width"
  );
  XCTAssertEqual(
    screenInfo.heightPixels,
    (NSUInteger)2532,
    @"heightPixels should match device type mainScreenSize.height"
  );
  XCTAssertEqualWithAccuracy(screenInfo.scale, 3.0f, 0.001f, @"scale should match device type mainScreenScale");
}

#pragma mark - Equality and Hashing

- (void)testIsEqual_WhenSameDevice_ReturnsYES
{
  FBSimulator *sim1 = _createSimulatorWithDevice(_stubDevice);
  FBSimulator *sim2 = _createSimulatorWithDevice(_stubDevice);
  XCTAssertEqualObjects(
    sim1,
    sim2,
    @"Two simulators wrapping the same device should be equal"
  );
}

- (void)testIsEqual_WhenDifferentDevice_ReturnsNO
{
  _FBTestSimDevice *otherDevice = [[_FBTestSimDevice alloc] init];
  FBSimulator *sim2 = _createSimulatorWithDevice(otherDevice);
  XCTAssertNotEqualObjects(
    _simulator,
    sim2,
    @"Two simulators wrapping different devices should not be equal"
  );
}

- (void)testIsEqual_WhenDifferentClass_ReturnsNO
{
  XCTAssertFalse(
    [_simulator isEqual:@"not a simulator"],
    @"isEqual should return NO for objects of a different class"
  );
}

- (void)testHash_MatchesDeviceHash
{
  NSUInteger expected = [_stubDevice hash];
  XCTAssertEqual(
    _simulator.hash,
    expected,
    @"Simulator hash should match the underlying device hash"
  );
}

#pragma mark - Temporary Directory (Lazy Init)

- (void)testTemporaryDirectory_ReturnsSameInstanceOnSubsequentAccess
{
  FBTemporaryDirectory *first = _simulator.temporaryDirectory;
  FBTemporaryDirectory *second = _simulator.temporaryDirectory;
  XCTAssertTrue(
    first == second,
    @"temporaryDirectory should return the same cached instance on subsequent access"
  );
}

#pragma mark - Healthcheck Helpers

- (void)testLookupBootstrapPortNamed_WhenPortExists_ReturnsPortNumber
{
  _stubDevice.lookupPort = 12345;
  _stubDevice.lookupShouldFail = NO;
  FBSimulator *sim = _createSimulatorWithDevice(_stubDevice);
  NSError *error = nil;
  NSNumber *port = [sim lookupBootstrapPortNamed:@"com.apple.testservice" error:&error];
  XCTAssertNil(error, @"Error should be nil when port lookup succeeds");
  XCTAssertNotNil(port, @"Port should not be nil when lookup succeeds");
  XCTAssertEqual(
    port.unsignedIntValue,
    12345u,
    @"Returned port number should match the looked-up port"
  );
}

- (void)testLookupBootstrapPortNamed_WhenPortIsNull_ReturnsNil
{
  _stubDevice.lookupPort = MACH_PORT_NULL;
  _stubDevice.lookupShouldFail = NO;
  FBSimulator *sim = _createSimulatorWithDevice(_stubDevice);
  NSError *error = nil;
  NSNumber *port = [sim lookupBootstrapPortNamed:@"com.apple.nonexistent" error:&error];
  XCTAssertNil(port, @"Port should be nil when MACH_PORT_NULL is returned");
}

- (void)testLookupBootstrapPortNamed_WhenLookupFails_ReturnsNilWithError
{
  _stubDevice.lookupShouldFail = YES;
  FBSimulator *sim = _createSimulatorWithDevice(_stubDevice);
  NSError *error = nil;
  NSNumber *port = [sim lookupBootstrapPortNamed:@"com.apple.failing" error:&error];
  XCTAssertNil(port, @"Port should be nil when lookup fails");
  XCTAssertNotNil(error, @"Error should be populated when lookup fails");
}

@end

File: FBSimulatorControlTests/Tests/Unit/FBSimulatorTests.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

@preconcurrency import CoreSimulator
import FBControlCore
@testable import FBSimulatorControl
import XCTest

// MARK: - Test Doubles
//
// The simulator reaches its device by Objective-C message send, having been handed one of these
// in place of a real `SimDevice`, so the doubles are `@objc` `NSObject` subclasses whose
// property names match the selectors the simulator sends.

@objc final class TestSimDeviceType: NSObject {
  @objc var productFamilyID: Int32 = 1
  @objc var mainScreenSize: CGSize = CGSize(width: 750, height: 1334)
  @objc var mainScreenScale: Float = 2.0
  @objc var name: String = "iPhone 14"
}

@objc final class TestSimRuntime: NSObject {
  @objc var root: String = "/Library/Developer/CoreSimulator/Profiles/Runtimes/iOS.simruntime/Contents/Resources/RuntimeRoot"
  @objc var name: String = "iOS 17.0"
}

@objc final class TestSimDeviceSet: NSObject {
  @objc var setPath: String = ("~/Library/Developer/CoreSimulator/Devices" as NSString).expandingTildeInPath
}

@objc final class TestSimDevice: NSObject {
  @objc var UDID: NSUUID = NSUUID()
  @objc var name: String = "TestSimulator"
  @objc var state: UInt = UInt(FBiOSTargetState.booted.rawValue)
  @objc var dataPath: String = "/tmp/test-sim-data"
  @objc var deviceType = TestSimDeviceType()
  @objc var runtime = TestSimRuntime()
  @objc var deviceSet = TestSimDeviceSet()
  @objc var lookupPort: mach_port_t = mach_port_t(MACH_PORT_NULL)
  @objc var lookupShouldFail = false

  @objc(lookup:error:)
  func lookup(_ name: String, error: NSErrorPointer) -> mach_port_t {
    if lookupShouldFail {
      error?.pointee = NSError(
        domain: "FBSimulatorTestDomain",
        code: 1,
        userInfo: [NSLocalizedDescriptionKey: "Port not found"])
      return mach_port_t(MACH_PORT_NULL)
    }
    return lookupPort
  }
}

// MARK: - Test Class

final class FBSimulatorTests: XCTestCase {

  private var stubDevice: TestSimDevice!
  private var simulator: FBSimulator!

  override func setUpWithError() throws {
    try super.setUpWithError()
    stubDevice = TestSimDevice()
    simulator = Self.createSimulator(with: stubDevice)
  }

  override func tearDownWithError() throws {
    simulator = nil
    stubDevice = nil
    try super.tearDownWithError()
  }

  private static func createSimulator(with device: TestSimDevice) -> FBSimulator {
    FBSimulatorTestSupport.testableSimulator(withDevice: device)
  }

  // MARK: - Product Family

  func testProductFamily_WhenFamilyIDIs1_ReturnsiPhone() {
    stubDevice.deviceType.productFamilyID = 1
    let sim = Self.createSimulator(with: stubDevice)
    XCTAssertEqual(sim.productFamily, .familyiPhone, "productFamilyID 1 should map to iPhone")
  }

  func testProductFamily_WhenFamilyIDIs2_ReturnsiPad() {
    stubDevice.deviceType.productFamilyID = 2
    let sim = Self.createSimulator(with: stubDevice)
    XCTAssertEqual(sim.productFamily, .familyiPad, "productFamilyID 2 should map to iPad")
  }

  func testProductFamily_WhenFamilyIDIs3_ReturnsAppleTV() {
    stubDevice.deviceType.productFamilyID = 3
    let sim = Self.createSimulator(with: stubDevice)
    XCTAssertEqual(sim.productFamily, .familyAppleTV, "productFamilyID 3 should map to AppleTV")
  }

  func testProductFamily_WhenFamilyIDIs4_ReturnsAppleWatch() {
    stubDevice.deviceType.productFamilyID = 4
    let sim = Self.createSimulator(with: stubDevice)
    XCTAssertEqual(sim.productFamily, .familyAppleWatch, "productFamilyID 4 should map to AppleWatch")
  }

  func testProductFamily_WhenFamilyIDIsUnknown_ReturnsUnknown() {
    stubDevice.deviceType.productFamilyID = 99
    let sim = Self.createSimulator(with: stubDevice)
    XCTAssertEqual(sim.productFamily, .familyUnknown, "Unrecognized productFamilyID should map to Unknown")
  }

  // MARK: - Custom Device Set Path

  func testCustomDeviceSetPath_WhenDefaultPath_ReturnsNil() {
    stubDevice.deviceSet.setPath = ("~/Library/Developer/CoreSimulator/Devices" as NSString).expandingTildeInPath
    let sim = Self.createSimulator(with: stubDevice)
    XCTAssertNil(sim.customDeviceSetPath, "customDeviceSetPath should be nil when using the default device set path")
  }

  func testCustomDeviceSetPath_WhenCustomPath_ReturnsPath() {
    let customPath = "/custom/simulator/devices"
    stubDevice.deviceSet.setPath = customPath
    let sim = Self.createSimulator(with: stubDevice)
    XCTAssertEqual(
      sim.customDeviceSetPath, customPath,
      "customDeviceSetPath should return the custom path when not using the default")
  }

  // MARK: - Screen Info

  func testScreenInfo_ReturnsDeviceTypeScreenDimensions() throws {
    stubDevice.deviceType.mainScreenSize = CGSize(width: 1170, height: 2532)
    stubDevice.deviceType.mainScreenScale = 3.0
    let sim = Self.createSimulator(with: stubDevice)

    let screenInfo = try XCTUnwrap(sim.screenInfo, "screenInfo should not be nil")

    XCTAssertEqual(screenInfo.widthPixels, 1170, "widthPixels should match device type mainScreenSize.width")
    XCTAssertEqual(screenInfo.heightPixels, 2532, "heightPixels should match device type mainScreenSize.height")
    XCTAssertEqual(screenInfo.scale, 3.0, accuracy: 0.001, "scale should match device type mainScreenScale")
  }

  // MARK: - Equality and Hashing

  func testIsEqual_WhenSameDevice_ReturnsYES() {
    let sim1 = Self.createSimulator(with: stubDevice)
    let sim2 = Self.createSimulator(with: stubDevice)
    XCTAssertEqual(sim1, sim2, "Two simulators wrapping the same device should be equal")
  }

  func testIsEqual_WhenDifferentDevice_ReturnsNO() {
    let sim2 = Self.createSimulator(with: TestSimDevice())
    XCTAssertNotEqual(simulator, sim2, "Two simulators wrapping different devices should not be equal")
  }

  func testHash_DelegatesToDevice() {
    let sim1 = Self.createSimulator(with: stubDevice)
    let sim2 = Self.createSimulator(with: stubDevice)
    XCTAssertEqual(sim1.hashValue, sim2.hashValue, "Simulators wrapping the same device should hash identically")
    XCTAssertEqual(Set([sim1, sim2]).count, 1, "A set should collapse simulators wrapping the same device")
  }

  // MARK: - Temporary Directory (Lazy Init)

  func testTemporaryDirectory_ReturnsSameInstanceOnSubsequentAccess() {
    let first = simulator.temporaryDirectory
    let second = simulator.temporaryDirectory
    XCTAssertTrue(first === second, "temporaryDirectory should return the same cached instance on subsequent access")
  }

  // MARK: - Healthcheck Helpers

  func testLookupBootstrapPortNamed_WhenPortExists_ReturnsPortNumber() throws {
    stubDevice.lookupPort = 12345
    stubDevice.lookupShouldFail = false
    let sim = Self.createSimulator(with: stubDevice)

    let port = try sim.lookupBootstrapPortNamed("com.apple.testservice")

    XCTAssertEqual(try XCTUnwrap(port).uint32Value, 12345, "Returned port number should match the looked-up port")
  }

  // Both no-port paths throw: the device's own error when it reported one, and a synthesized
  // `FBSimulatorPortLookupError` when it did not. They are told apart by the error's domain.
  func testLookupBootstrapPortNamed_WhenPortIsNull_ProducesNoPort() {
    stubDevice.lookupPort = mach_port_t(MACH_PORT_NULL)
    stubDevice.lookupShouldFail = false
    let sim = Self.createSimulator(with: stubDevice)

    XCTAssertThrowsError(try sim.lookupBootstrapPortNamed("com.apple.nonexistent"), "No port should be produced when MACH_PORT_NULL is returned") { error in
      XCTAssertNotEqual(
        (error as NSError).domain, "FBSimulatorTestDomain",
        "A null port is not a lookup failure - the device reported no error of its own")
    }
  }

  func testLookupBootstrapPortNamed_WhenLookupFails_ReturnsNilWithError() {
    stubDevice.lookupShouldFail = true
    let sim = Self.createSimulator(with: stubDevice)

    XCTAssertThrowsError(try sim.lookupBootstrapPortNamed("com.apple.failing"), "Error should be populated when lookup fails") { error in
      XCTAssertEqual((error as NSError).domain, "FBSimulatorTestDomain", "The device's own lookup error should surface")
    }
  }
}
